File: lib/services/docker.dart
library dork.docker;

import "dart:async";
import "dart:io";
import "dart:convert";
import "package:dork/services/service.dart";
import "package:dork/services/environment.dart";
import "package:http/http.dart" as http;

/// Simple struct representing a Docker container.
class Container {
  String id;
  String image;
  String name;
  String address;
  String workDirectory;
  String buildDirectory;

  bool get running => address != null;
  operator == (Container item) {
    if (item.id != null && item.id != this.id) return false;
    if (item.image != null && item.image != this.image) return false;
    if (item.name != null && item.name != this.name) return false;
    if (item.address != null && item.address != this.address) return false;
    if (item.workDirectory != null && item.workDirectory != this.workDirectory) return false;
    if (item.buildDirectory != null && item.buildDirectory != this.buildDirectory) return false;
    return true;
  }

  String get hash {
    String hash = this.name.split('.').last;
    if (hash == 'new') hash = null;
    return hash;
  }

  String get project => this.name.split('.').first;

  String get instance => this.name.split('.').elementAt(1);

  static String naming(String project, String instance, String hash) => "${project}.${instance}.${hash}";

  String toString() {
    return JSON.encode({
      'id': this.id,
      'image': this.image,
      'name': this.name,
      'address': this.address,
      'workDirectory': this.workDirectory,
      'buildDirectory': this.buildDirectory,
    });
  }
}

/// Simple struct representing a Docker image.
class Image {
  String id;
  String name;
  operator == (Image item) {
    if (item.id != null && item.id != this.id) return false;
    if (item.name != null && item.name != this.name) return false;
    return true;
  }

  String get project => this.name.split('/').first;
  String get hash => this.name.split('/').last;
  static String naming(String project, String hash) => "${project}/${hash}";

  String toString() {
    return JSON.encode({
      'id': this.id,
      'name': this.name
    });
  }
}

class RawDocker extends Service {

  final Environment env;

  RawDocker(this.env);

  Future writeHosts(Map<String, String> hosts) {
    List<String> file = [];
    bool dork_hosts = false;
    bool writter = false;
    File hostsfile = new File('/etc/hosts');
    hostsfile.readAsLinesSync().forEach((String line){
      if (line == "# DORK START") dork_hosts = true;
      if (!dork_hosts) file.add(line);
      if (line == "# DORK END") dork_hosts == false;
    });
    if (hosts.length > 0) {
      file.add('# DORK START');
      hosts.forEach((String ip, String domain) {
        file.add("${ip} ${domain}");
      });
      file.add('# DORK END');
    }
    hostsfile.writeAsString(file.join('\n'));
    this.run('sudo', ['service', 'dnsmasq', 'restart']);
  }

  Future<dynamic> _apiGet(String url, [ List<int> accepted = Service.default_accepted ]) async {
    return JSON.decode(await this.get(this.env.dockerAddress + '/' + url, accepted));
  }

  Future<dynamic> _apiPost(String url, [ Map params = Service.default_params, List<int> accepted = Service.default_accepted ]) async {
    String req = this.env.dockerAddress + url;
    String result = await this.post(req, params, accepted);
    try {
      return JSON.decode(result);
    } catch (exception) {
      return null;
    }
  }

  Future<dynamic> _apiDelete(String url, [ List<int> accepted = Service.default_accepted ]) async {
    String req = this.env.dockerAddress + url;
    String result = await this.delete(req, accepted);
    try {
      return JSON.decode(result);
    } catch (exception) {
      return null;
    }
  }

  Future<List<Container>> getContainers() async {
    List<Container> containers = [];
    await Future.forEach(await this._apiGet('/containers/json?all=1'), (Map data) async {
      Map c = await this._apiGet("/containers/${data['Id']}/json");
      String id = c['Id'];
      String name = c['Name'];
      if (name.substring(0, 1) == '/') {
        name = name.substring(1);
      }
      String image = c['Image'];
      String address = c['NetworkSettings']['IPAddress'];
      bool running = c['State']['Running'];
      String workdir = null;
      String builddir = null;
      c['HostConfig']['Binds'].forEach((String bind) {
        try {
          String host = bind.split(':').first;
          String container = bind.split(':').last;
          if (container == this.env.dorkBuildDirectory) {
            builddir = host;
          }
          if (container == this.env.dorkSourceDirectory) {
            workdir = host;
          }
        } catch (exception) {}
      });
      if (workdir != null && builddir != null) {
        containers.add(new Container()
          ..id = id
          ..name = name
          ..image = image
          ..address = running ? address : null
          ..workDirectory = workdir
          ..buildDirectory = builddir
        );
      }
    });
    return containers;
  }

  Future<List<Image>> getImages() async {
    List<Image> images = [];
    (await this._apiGet('/images/json')).forEach((Map i) {
      if (!i.containsKey('RepoTags') || i['RepoTags'].length == 0 ) {
        return;
      }
      String name = i['RepoTags'][0].split(':').first;
      String id = i['Id'];
      images.add(new Image()
        ..id = id
        ..name = name
      );
    });
    return images;
  }

  Future<List<String>> getDanglingImages() async {
    return this.run('docker', ['images', '-q', '-f', 'dangling=true']);
  }

  Future create(String name, String image, Map<String, String> volumes) async {
    Map conf = {
      'Image': image,
      'Volumes': {},
      'HostConfig': {
        'Binds': []
      },
    };
    volumes.forEach((String host, String container) {
      conf['Volumes'][container] = {};
      conf['HostConfig']['Binds'].add("${host}:${container}");
    });
    String url = '/containers/create?name=' + name;
    await this._apiPost(url , conf, [201]);
  }

  Future start(String container) async {
    await this._apiPost('/containers/${container}/start', null, [204, 304]);
  }

  Future stop(String container) async {
    await this._apiPost('/containers/${container}/stop', null, [204, 304]);
  }

  Future remove(String container) async {
    await this._apiDelete('/containers/${container}', [204]);
  }

  Future removeImage(String image) async {
    await this._apiDelete('/images/${image}', [200]);
  }

  Future rename(String container, String name) async {
    await this._apiPost('/containers/${container}/rename?name=${name}', null, [204]);
  }

  Future commit(String container, String repository) async {
    await this._apiPost('/commit?container=${container}&repo=${repository}', {}, [201]);
  }

  Future<bool> isAccessible(String container) async {
    Map params = {
      "AttachStdin": false,
      "AttachStdout": true,
      "AttachStderr": true,
      "Tty": false,
      "Cmd": [ "ps -A | grep sshd" ],
    };
    Map exec = await this._apiPost('/containers/${container}/exec', params, [201]);
    String result = await this.post(this.env.dockerAddress + '/exec/${exec['Id']}/start', { 'Detach': false, "Tty": false }, [200]);
    return result.contains('sshd');
  }
}

abstract class Docker implements RawDocker {
  Future<Container> containerById(String id) async {
    Container c = (await this.getContainers()).firstWhere((Container c) => c.id == id, orElse: () => null);
    if (c == null) throw new StateError('Container with id ${id} doesn\'t exist.');
    return c;
  }

  Future<Container> containerByName(String name) async {
    Container c = (await this.getContainers()).firstWhere((Container c) => c.name == name, orElse: () => null);
    if (c == null) throw new StateError('Container with name ${name} doesn\'t exist.');
    return c;
  }

  Future<Image> imageById(String id) async {
    Image i = (await this.getImages()).firstWhere((Image i) => i.id == id, orElse: () => null);
    if (i == null) throw new StateError('Image with id ${id} doesn\'t exist.');
    return i;
  }

  Future<Image> imageByName(String name) async {
    Image i = (await this.getImages()).firstWhere((Image i) => i.name == name, orElse: () => null);
    if (i == null) throw new StateError('Image with name ${name} doesn\'t exist.');
    return i;
  }

  Future updateHosts() async{
    Map<String, String> hosts = {};
    (await this.getContainers()).forEach((Container c) {
      String project = c.name.split('.').elementAt(0);
      String instance = c.name.split('.').elementAt(1);
      hosts[c.address] = "${project}.${instance}.dork";
      if (instance == project) {
        hosts[c.address] += " ${project}.dork";
      }
    });
    this.writeHosts(hosts);
  }
}

File: dork/docker.py
"""
Simple API to docker.
"""

import requests
from git import Repository
from config import config
from subprocess import check_output, call, Popen, PIPE
from dateutil.parser import parse as parse_date
import json
import os
import rx
import rx.subjects
import threading
import re
from rx import Observable


def __eventstream(stream, killsignal):
    process = Popen('docker events', stdout=PIPE, shell=True)
    killsignal.subscribe(lambda v: process.kill())
    for line in iter(process.stdout.readline, ''):
        stream.on_next(line)

eventpattern = re.compile('(.*?) (.*):.*?([a-z]*)$')


def __parseevent(line):
    return {
        'timestamp': eventpattern.match(line).group(1),
        'id': eventpattern.match(line).group(2),
        'event': eventpattern.match(line).group(3),
    }


def __event_object(event):
    for c in containers(True):
        if c.id == event['id']:
            event['container'] = c
    for i in images(True):
        if i.id == event['id']:
            event['image'] = i
    return event


_docker_events = None


def events(killsignal):
    """
    :rtype: Observable
    """
    global _docker_events
    if not _docker_events:
        _docker_events = rx.subjects.Subject()
        thread = threading.Thread(target=__eventstream, args=(_docker_events, killsignal))
        thread.start()
    return _docker_events.map(__parseevent).map(__event_object)


class Container:
    """
    Class representing a container, providing the necessary information and
    operations.
    """
    def __init__(self, data):
        """
        :param dict data: dict data: The dataset returned by the Docker API.
        """
        self.__data = data

    def __str__(self):
        return self.name

    @classmethod
    def list(cls, clear=False):
        return containers(clear)

    @classmethod
    def create(cls, name, image, volumes, hostname):
        return create(name, image, volumes, hostname)

    def export(self, filename):
        """
        Export container to a file.
        """
        call('docker export %s > %s' % (self.id, filename), shell=True, stdout=open(os.devnull, 'w'))

    @property
    def id(self):
        """
        The containers Id.

        :rtype: str
        """
        return self.__data['Id']

    @property
    def image(self):
        """
        The image's id the container was created from.

        :rtype: str
        """
        return self.__data['Image']

    @property
    def name(self):
        """
        The containers name.

        :rtype: str
        """
        return self.__data['Name']

    @property
    def project(self):
        """
        The containers project. First segment of [Container.name].

        :rtype: str
        """
        return self.name.split('.')[0].strip('/')

    @property
    def instance(self):
        """
        The containers instance. Second segment of [Container.name].

        :rtype: str
        """
        return self.name.split('.')[1]

    @property
    def hash(self):
        """
        The containers git hash. Second segment of [Container.name].

        :rtype: str
        """
        return self.name.split('.')[2]

    @property
    def domain(self):
        """
        The containers internal domain name.

        :rtype: str
        """
        if self.project == self.instance:
            return "%s.dork" % self.project
        else:
            return "%s.%s.dork" % (self.project, self.instance)

    @property
    def running(self):
        """
        Check if the container is running.

        :rtype: bool
        """
        return self.__data['State']['Running']

    @property
    def address(self):
        """
        The containers IP address, or [None] if the container is not running.

        :rtype: str
        """
        if self.running:
            return self.__data['NetworkSettings']['IPAddress']
        else:
            return None

    @property
    def source(self):
        """
        The directory on the host machine, mounted to the containers source
        directory.

        :rtype: str
        """
        directory = None
        for bind in self.__data['HostConfig']['Binds']:
            host = bind.split(':')[0]
            container = bind.split(':')[1]
            if container == config.dork_source_directory:
                directory = host
        return directory

    @property
    def repository(self):
        """
        Retrieve the containers source git repository.

        :rtype: Repository
        """
        return Repository(self.source)

    @property
    def build(self):
        """
        The directory on the host machine, mounted to the containers build
        directory.

        :rtype: str
        """
        directory = None
        for bind in self.__data['HostConfig']['Binds']:
            host = bind.split(':')[0]
            container = bind.split(':')[1]
            if container == config.dork_build_directory:
                directory = host
        return directory

    @property
    def logs(self):
        """
        The directory on the host machine, mounted to the containers build
        directory.

        :rtype: str
        """
        directory = None
        for bind in self.__data['HostConfig']['Binds']:
            host = bind.split(':')[0]
            container = bind.split(':')[1]
            if container == config.dork_log_directory:
                directory = host
        return directory

    @property
    def accessible(self):
        """:rtype: bool"""
        if not self.running:
            return False
        else:
            return _container_accessible(self.address)

    @property
    def time_created(self):
        """:rtype: datetime"""
        return parse_date(self.__data['Created'])

    @property
    def time_started(self):
        """:rtype: datetime"""
        if self.running:
            return parse_date(self.__data['State']['StartedAt'])
        else:
            return None

    @property
    def time_stopped(self):
        """:rtype: datetime"""
        if self.running:
            return None
        else:
            return parse_date(self.__data['State']['FinishedAt'])

    def hostPort(self, port):
        return self.__data['NetworkSettings']['Ports']['80/tcp'][0]['HostPort'] if '80/tcp' in self.__data['NetworkSettings']['Ports'] else None

    def start(self):
        _container_start(self.id)

    def stop(self):
        _container_stop(self.id)

    def remove(self):
        _container_remove(self.id)

    def rename(self, name):
        _container_rename(self.id, name)

    def commit(self, repo):
        _container_commit(self.id, repo)

    def execute(self, command):
        _container_execute(self.id, command)


class Image:
    def __init__(self, data):
        self.__data = data

    def __str__(self):
        return self.name

    @classmethod
    def list(cls, clear=False):
        return images(clear)

    @classmethod
    def fromFile(cls, file, name):
        call('cat %s | docker import - %s' % (file, name), shell=True, stdout=open(os.devnull, 'w'))
        images(True)

    @classmethod
    def dangling(cls):
        return _dangling_images()

    @property
    def id(self):
        """:rtype: str"""
        return self.__data['Id']

    @property
    def name(self):
        """:rtype: str"""
        return self.__data['RepoTags'][0].split(':')[0]

    @property
    def project(self):
        """:rtype: str"""
        return self.name.split('/')[0]

    @property
    def hash(self):
        """:rtype: str"""
        return self.name.split('/')[1]

    @property
    def time_created(self):
        """:rtype: datetime"""
        return parse_date(self.__data['Created'])

    def delete(self):
        _image_remove(self.id)


class BaseImage:
    def __init__(self, project, base):
        self.project = project
        self.name = base
        self.hash = 'new'


# ======================================================================
# PUBLIC METHODS
# ======================================================================
__containers = None
def containers(clear=False):
    global __containers
    if __containers is None or clear:
        __containers = []
        for cid in check_output(['docker', 'ps', '-aq']).splitlines():
            data = json.loads(check_output(['docker', 'inspect', cid]))
            __containers.append(Container(data[0]))
    return __containers


__images = None
def images(clear=False):
    global __images
    if __images is None or clear:
        __images = []
        for i in check_output(['docker', 'images', '-q']).splitlines():
            data = json.loads(check_output(['docker', 'inspect', i]))
            if (data[0]['RepoTags']):
                __images.append(Image(data[0]))
    return __images


def create(name, image, volumes, hostname):
    """:type volumes: dict"""
    cmd = ['docker', 'create', '--name=%s' % name, '-h', hostname, '-P']
    for host in volumes:
        cmd.append('-v')
        cmd.append("%s:%s" % (host, volumes[host]))

    cmd.append(image)
    cmd.append('/usr/bin/supervisord')
    check_output(cmd)
    containers(True)


def _dangling_images():
    image_ids = check_output([
        'docker', 'images', '-q', '-f', 'dangling=true'
    ]).splitlines()
    for iid in image_ids:
        yield Image(json.loads(check_output(['docker', 'inspect', iid]))[0])


# ======================================================================
# PROTECTED METHODS
# ======================================================================
def _container_start(cid):
    check_output(['docker', 'start', cid])
    containers(True)


def _container_stop(cid):
    check_output(['docker', 'stop', cid])
    containers(True)


def _container_remove(cid):
    check_output(['docker', 'rm', cid])
    containers(True)


def _container_rename(cid, name):
    check_output(['docker', 'rename', cid, name])
    containers(True)


def _container_commit(cid, repo):
    check_output(['docker', 'commit', cid, repo])
    images(True)


def _container_accessible(address):
    return call(['ssh', '-F', os.path.expanduser('~/.ssh/config'), address, '/bin/true']) == 0

def _container_execute(id, command):
    check_output(['docker', 'exec', id, command])



def _image_remove(iid):
    call(['docker', 'rmi', iid])
    images(True)


# ======================================================================
# PRIVATE METHODS
# ======================================================================
class DockerException(Exception):
    def __init__(self, msg, code):
        super(DockerException, self).__init__(msg)
        self.code = code


def __get(path, query=(), codes=(200,)):
    """
    :param str path:
    :param dict[str,str] query:
    :param list[int] codes:
    :return: json
    """

    result = requests.get(
        "%s/%s" % (config.docker_address, path),
        params=query)

    if result.status_code in codes:
        return result.json()
    else:
        raise DockerException(result.status_code, result.text)


def __post(path, query=(), data=(), codes=(200,)):

    if data:
        result = requests.post(
            "%s/%s" % (config.docker_address, path),
            params=query, json=data)
    else:
        result = requests.post(
            "%s/%s" % (config.docker_address, path),
            params=query)

    if result.status_code not in codes:
        raise DockerException(result.status_code, result.text)
    return result.text


def __delete(path, query=(), codes=(200,)):

    result = requests.delete(
        "%s/%s" % (config.docker_address, path),
        params=query)

    if result.status_code not in codes:
        raise DockerException(result.status_code, result.text)
